application_handlers_test.go

```diff
@@ -854,3 +856,114 @@ func applicationEventTestHelper(t *testing.T, c echo.Context, expectedEventType
 
 	return nil
 }
+
+// TestApplicationEditPausedIntegration tests that a  "bad request" response is returned when a paused application is
+// tried to be updated. Runs on integration tests.
+func TestApplicationEditPausedIntegration(t *testing.T) {
+	testutils.SkipIfNotRunningIntegrationTests(t)
+
+	req := m.ApplicationEditRequest{
+		Extra:                   []byte(`{"thing": true}`),
+		AvailabilityStatus:      request.PointerToString("available"),
+		AvailabilityStatusError: request.PointerToString(""),
+	}
+
+	body, _ := json.Marshal(req)
+
+	c, rec := request.CreateTestContext(
+		http.MethodPatch,
+		"/api/sources/v3.1/applications/1",
+		bytes.NewReader(body),
+		map[string]interface{}{
+			"tenantID": int64(1),
+		},
+	)
+
+	c.SetParamNames("id")
+	c.SetParamValues("1")
+	c.Request().Header.Add("Content-Type", "application/json;charset=utf-8")
+
+	// Modify the application so that the underlying code identifies it as "paused".
+	err := dao.DB.Model(m.Application{}).Where("id = ?", 1).UpdateColumn("paused_at", time.Now()).Error
+	if err != nil {
+		t.Error(err)
+	}
+
+	badRequestApplicationEdit := ErrorHandlingContext(ApplicationEdit)
+	err = badRequestApplicationEdit(c)
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Revert the changes so other tests don't have any problems.
+	err = dao.DB.Model(m.Application{}).Where("id = ?", 1).UpdateColumn("paused_at", nil).Error
+	if err != nil {
+		t.Error(err)
+	}
+
+	if rec.Code != http.StatusBadRequest {
+		t.Errorf("Wrong return code, expected %v got %v", http.StatusBadRequest, rec.Code)
+	}
+
+	want := "can't modify a paused application. Unpause it first and try again"
+	got := rec.Body.String()
+	if !strings.Contains(got, want) {
+		t.Errorf(`unexpected body returned. Want "%s" contained in what we got "%s"`, want, got)
+	}
+}
+
+// TestApplicationEditPausedUnit tests that a "bad request" response is returned when a paused application is tried to
+// be updated. Runs on unit tests.
```

Comment:
I am not sure about "Runs on unit tests". Does this mean that this test is not running with integration tests? I am asking because I cannot see here something like `testutils.SkipIfNotRunningIntegrationTests(t)`.

Reply:
I rephrased those comments and hopefully are more clear now.

Reply:
🧐. ... so this test is running with unit (mocked) test and with integration test, but in both cases the test handle the job as unit (mocked) test?

Reply:
Exactly, yes. I did this to avoid having something like this in the tests:

```go
var backupDao func(c echo.Context) (dao.SourceDao, error)
if parser.RunningIntegrationTests {
	// Modify the source so that the underlying code identifies it as "paused".
	err := dao.DB.Model(m.Source{}).Where("id = ?", 1).UpdateColumn("paused_at", time.Now()).Error
	if err != nil {
		t.Error(err)
	}
} else {
	// Get the specific SourceDao mock which simulates that the sources are paused.
	backupDao = getSourceDao
	getSourceDao = func(c echo.Context) (dao.SourceDao, error) {
		return &dao.MockSourceDao{IsPausedReturnValue: true}, nil
	}

}

badRequestSourceEdit := ErrorHandlingContext(SourceEdit)
err := badRequestSourceEdit(c)
if err != nil {
	t.Error(err)
}

if parser.RunningIntegrationTests {
	// Revert the changes so other tests don't have any problems.
	err = dao.DB.Model(m.Source{}).Where("id = ?", 1).UpdateColumn("paused_at", nil).Error
	if err != nil {
		t.Error(err)
	}
} else {
	// Go back to the previous DAO mock.
	getSourceDao = backupDao
}
```

Since all the integration tests begin with the `testutils.SkipIfNotRunningIntegrationTests(t)` I thought of following the same pattern at the expense of writing a "kind-of-duplicated" tests for this feature. Just for consistency.